pkg/apps/strategy/recreate/recreate.go

```diff
@@ -253,38 +252,45 @@ func (s *RecreateDeploymentStrategy) scaleAndWait(deployment *kapi.ReplicationCo
 	return s.rcClient.ReplicationControllers(deployment.Namespace).Get(deployment.Name, metav1.GetOptions{})
 }
 
-// waitForTerminatedPods waits until all pods for the provided replication controller are terminated.
-func (s *RecreateDeploymentStrategy) waitForTerminatedPods(from *kapi.ReplicationController, timeout time.Duration) {
-	selector := labels.SelectorFromValidatedSet(labels.Set(from.Spec.Selector))
-	options := metav1.ListOptions{LabelSelector: selector.String()}
-	podList, err := s.podClient.Pods(from.Namespace).List(options)
-	if err != nil {
-		fmt.Fprintf(s.out, "--> Cannot list pods: %v\nNew pods may be scaled up before old pods terminate\n", err)
-		return
-	}
-	// If there are no pods left, we are done.
-	if len(podList.Items) == 0 {
-		return
-	}
-	// Watch from the resource version of the list and wait for all pods to be deleted
-	// before proceeding with the Recreate strategy.
-	options.ResourceVersion = podList.ResourceVersion
-	w, err := s.podClient.Pods(from.Namespace).Watch(options)
-	if err != nil {
-		fmt.Fprintf(s.out, "--> Watch could not be established: %v\nNew pods may be scaled up before old pods terminate\n", err)
-		return
-	}
-	defer w.Stop()
-	// Observe as many deletions as the remaining pods and then return.
-	deletionsNeeded := len(podList.Items)
-	condition := func(event watch.Event) (bool, error) {
-		if event.Type == watch.Deleted {
-			deletionsNeeded--
+// oldPodsRunning returns whether there are old pods running or any of the old ReplicaSets thinks that it runs pods.
+func hasRunningPod(pods []kapi.Pod) bool {
+	for _, pod := range pods {
+		switch pod.Status.Phase {
+		case kapi.PodFailed, kapi.PodSucceeded:
```

Comment:
this also depends on kubelet not moving the state of pods right? probably worth adding a comment on that here for future

Reply:
I think seth has finally done the check also in apiserver, so that should be fixed. Not sure I want to add comments saying that kubernetes invariant **could get broken one day** if it's not broken now.